---
title: "Working with DNA motifs"
author: "Pierre-Luc"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(GenomicRanges)
  library(ggplot2)
  library(memes) # for the meme-based methods -- COMMENT OUT when using alternatives
  library(motifmatchr) # for scanning sequences for matches of given motifs
  library(Biostrings) # for handling sequences
  library(MotifDb) # database of motifs
  library(TFBSTools) # for handling some motif formats
  library(universalmotif) # for converting motifs to various formats
  library(PWMEnrich) # for R-based motif enrichment analysis
})
```


# Obtaining peak sequences

For the purpose of this example, we'll use the CTCF peaks found on chromosome 1 of mESC:

Download:

```{r}
download.file("https://www.encodeproject.org/files/ENCFF508CKL/@@download/ENCFF508CKL.bed.gz", "mESC_CTCF_ENCFF508CKL.bed.gz")
peaks <- rtracklayer::import("mESC_CTCF_ENCFF508CKL.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "Ensembl"  # to change the convention of the chromosome names to ensembl (i.e. without 'chr')
peaks_chr1 <- peaks[seqnames(peaks)=="1"]
```

We also get the genome sequence:

```{r}
ah <- AnnotationHub()
genome <- ah[["AH68356"]]
# we'll load it into memory:
genome_seqs <- import(genome)
# if your genome was a fasta file instead, import using Biostrings::readDNAStringSet()
```


# Motif scanning

Motif **scanning** aims at finding the _occurrences_ of a known motif (i.e motif 'matches') in a set of sequences.
As this is methodologically fairly simple, it doesn't really matter what method one uses, though results may differ because of how the thresholds are set (i.e. what is similar enough to the motif).

## Getting the desired motif

```{r}
# we search for "CTCF" in the motif database
motifs <- query(MotifDb, "CTCF")
# there are several matching motifs:
names(motifs)
# we select one:
motif <- motifs[["Mmusculus-HOCOMOCOv10-CTCF_MOUSE.H10MO.A"]]
motif
# we visualize it:
view_motifs(motifs[1:2])
```


## Using motifmatchr:

```{r}
# if you don't already have the genome in fasta format saved somewhere, convert it to that format:
Biostrings::writeXStringSet(genome_seqs, "genome.fa")
# we also need to convert the motif to a format that this package will accept
motif2 <- convert_motifs(motif, class="TFBSTools-PWMatrix")
# if we had a list of motifs instead, we'd convert them like this:
# motifs2 <- setNames(do.call(PWMatrixList, convert_motifs(motifs, class="TFBSTools-PWMatrix")), names(motifs))
moi <- motifmatchr::matchMotifs(motif2, subject=peaks_chr1, genome=Rsamtools::FaFile("genome.fa"),
                                out="positions") # specify that we want the function to return the genomic positions of each motif match
moi <- moi[[1]] # we scanned for just one motif, so we get the results for that motif
head(moi)
```

Then we could for instance ask questions such as how many of the peaks have a motif?

```{r}
table(overlapsAny(peaks_chr1, moi))
```


## Scanning the whole genome

To scan the whole genome, one needs to use the genome sequence as input subject.
For example, using `motifmatchr::matchMotif` :

```{r}
# run the scan:
motif_across_genome <- matchMotifs(motif2, subject=genome_seqs, out="positions")[[1]]
# when inputing matchMotifs directly with sequences, one has to do the following 
# to transform the output to GRanges:
names(motif_across_genome) <- names(genome_seqs)
motif_across_genome <- as(motif_across_genome, "GRanges")
head(motif_across_genome)
```




# Motif enrichment analysis

Motif **enrichment analysis** aims at finding _known_ motifs that are _enriched_ in a set of sequences (e.g. peaks) versus a background.
By default, the background is generated by shuffling the input sequences, but it can also be specified.

## Preparing the sequences 

Since the motif of a transcription factor is typically found towards the center or summit of the peak, it is common practice to look for motifs around the center of peaks when doing motif discovery. 
How far around depends on the resolution of your data, i.e. how narrow are your peaks (can be between 50bp to 500bp around), as well as the nature of the signal from which the peaks were obtained.
The first thing we do, then, is to extract the regions around the peaks, and then obtain the sequence from those regions:

```{r}
quantile(width(peaks_chr1))
peak_centers <- resize(peaks_chr1, fix="center", width=100)
quantile(width(peak_centers))
# we get the sequences corresponding to the peak centers:
peak_seqs <- memes::get_sequence(peak_centers, genome)
# or without the memes package:
peak_seqs <- Biostrings::getSeq(genome, peak_centers)
# names(peak_seqs) <- as.character(granges(peak_centers))
peak_seqs
as.character(peak_seqs[1])
```


## Example using AME from the Meme suite

```{r}
# obtain the motifs (e.g. here use all the mouse motifs)
# To avoid having redundant motifs, we query motifs from a single source:
motifs <- query(MotifDb, c("Mmusculus", "HOCOMOCOv10"))
ame <- memes::runAme(peak_seqs, database=convert_motifs(motifs), meme_path="~/meme/bin")
head(ame)
```

We could also plot the results:

```{r}
ggplot(ame, aes(log2((1+tp)/(1+fp)), -log10(adj.pvalue), size=tp_percent)) + 
  geom_point(alpha=0.3) + geom_text(aes(label=motif_id), data=head(ame)) +
  labs(x="log2(fold-enrichment)", size="% of set")
```


## Example using PWMEnrich (much slower, but doesn't require memes)

```{r, eval=FALSE}
# prepare multithreading (works only on mac/linux):
registerCoresPWMEnrich(4)
# launch the analysis:
res <- PWMEnrich::motifEnrichment(peak_seqs, convert_motifs(motifs, "PWMEnrich-PWM"))
# visualize:
groupReport(res)
```





# Motif discovery

Motif **discovery** aims at finding _new motifs_ that are enriched in a set of sequences (e.g. peaks) versus a background (e.g. random or control sequences). This is the step where different methods will differ the most.

## Example using Meme

```{r, eval=FALSE}
# for the sake of example, we'll just use the first 100 sequences since this is computationally demanding:
peak_seqs_subset <- head(peak_seqs,100)
# this requires having the meme suite installed
mr <- memes::runMeme( peak_seqs_subset, parse_genomic_coord = FALSE,
                      p=4, # use 4 threads
                      nmotifs=5, # determines how many motifs Meme will try to find
                      meme_path = "~/meme/bin/") # depending on your setup, you might need to specify where the meme binaries were installed

# we could also specify a background using the `neg` argument, i.e. instead of 
# looking for motifs against a random background, you can specify control (i.e. non-peak) sequences
```

We could look at the top 3 motifs:

```{r}
view_motifs(to_list(head(mr,3)))
```
and compare the top motif to the CTCF motif we extracted from the database:

```{r}
view_motifs(list(convert_motifs(motif), mr$motif[[1]]))
```




## Alterantive example using rGADEM (doesn't require installing the meme suite)

```{r, eval=FALSE}
library(rGADEM)
gadem <- rGADEM::GADEM(peak_seqs_subset, genome=genome, nmotifs=3)
view_motifs(motif)
view_motifs(rGADEM::getPWM(gadem[[1]]))
```


